core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -1003,9 +1003,17 @@ class LogManager(logDirs: Seq[File],
   /**
    * Map of log dir to logs by topic and partitions in that dir
    */
-  private def logsByDir: Map[String, Map[TopicPartition, Log]] = {
-    (this.currentLogs.toList ++ this.futureLogs.toList).toMap
-      .groupBy { case (_, log) => log.parentDir }
+  def logsByDir: Map[String, Map[TopicPartition, Log]] = {
+    // This code is called often by checkpoint processes and is written in a way that reduces
+    // allocations and CPU with many topic partitions.
+    // When changing this code please measure the changes with org.apache.kafka.jmh.server.CheckpointBench
+    val byDir = new mutable.AnyRefMap[String, mutable.AnyRefMap[TopicPartition, Log]]()
+    def addToDir(tp: TopicPartition, log: Log): Unit = {
+      byDir.getOrElseUpdate(log.parentDir, new mutable.AnyRefMap[TopicPartition, Log]()).put(tp, log)
+    }
+    currentLogs.foreach { case (tp, log) => addToDir(tp, log) }
+    futureLogs.foreach { case (tp, log) => addToDir(tp, log) }
```

Comment:
`scala.collection.Map` has a `def foreachEntry[U](f: (K, V) => U): Unit` method that avoids allocating tuples (it's a function with two parameters). If we want to avoid allocations, we should introduce one such method to `Pool` and use it here.

Reply:
I tried out your suggestion and interestingly saw more allocations. Note extreme use of the `var item` there in case there was an extra allocation there.

```
--- a/core/src/main/scala/kafka/log/LogManager.scala
+++ b/core/src/main/scala/kafka/log/LogManager.scala
@@ -1011,8 +1011,8 @@ class LogManager(logDirs: Seq[File],
     def addToDir(tp: TopicPartition, log: Log): Unit = {
       byDir.getOrElseUpdate(log.parentDir, new mutable.AnyRefMap[TopicPartition, Log]()).put(tp, log)
     }
-    currentLogs.foreach { case (tp, log) => addToDir(tp, log) }
-    futureLogs.foreach { case (tp, log) => addToDir(tp, log) }
+    currentLogs.foreachEntry(addToDir)
+    futureLogs.foreachEntry(addToDir)
     byDir
   }
 
diff --git a/core/src/main/scala/kafka/utils/Pool.scala b/core/src/main/scala/kafka/utils/Pool.scala
index 964de7eae2..450f90611b 100644
--- a/core/src/main/scala/kafka/utils/Pool.scala
+++ b/core/src/main/scala/kafka/utils/Pool.scala
@@ -74,6 +74,15 @@ class Pool[K,V](valueFactory: Option[K => V] = None) extends Iterable[(K, V)] {
   def values: Iterable[V] = pool.values.asScala
 
   def clear(): Unit = { pool.clear() }
+
+  def foreachEntry(f: (K, V) => Unit): Unit = {
+    val iter = iterator
+    var item: (K, V) = null
+    while(iter.hasNext) {
+      item = iter.next()
+      f(item._1, item._2)
+    }
+  }

CheckpointBench.measureCheckpointLogStartOffsets:·gc.alloc.rate.norm                         3         2000  thrpt   15  1428889.965 ±  75131.113    B/op
vs trunk:
CheckpointBench.measureCheckpointLogStartOffsets:·gc.alloc.rate.norm                         3         2000  thrpt   15  1284326.850 ±  75148.430    B/op

```

Reply:
^ updated text there to say "this PR" rather than trunk.

Reply:
You have to use the iterator of `ConcurrentHashMap[K, V]` which gives you `Map.Entry` instead of a tuple. Then you avoid it.

Reply:
(or `forEach` in the `ConcurrentHashMap`)

Reply:
Duh, I was a bit hasty on that one. The fact that it was still a tuple was a dead give away...

New results are better, 18% reduction:
```
CheckpointBench.measureCheckpointLogStartOffsets:·gc.alloc.rate.norm                         3         2000  thrpt   15  1044120.722 ±  1676.137    B/op
```

Reply:
Do you know what I'm doing wrong when I supply a lambda to the java `.forEach`? I know it works for consumers but in my brief effort improving this I couldn't make it work with the BiConsumer here and had to create a new one.

I will update the body with the rerun jmh results.

Reply:
What error are you getting?

Reply:
AFAICT, it easy enough to do something like this:
```
pool.forEach((k,v) => f(k,v))
```
But then we appear to be back to allocating tuples.

It doesn't appear easy to do something like:
```
pool.forEach(kv => f(kv.getKey, kv.getValue))
```
since it's not able to create a BiConsumer for you.

Reply:
Why do you say that the following would allocate tuples?

```scala
def foreachEntry(f: (K, V) => Unit): Unit = {
    pool.forEach((k, v) => f(k, v))
  }
```

Did you measure it? Looking at the code, I don't see any tuples.

Reply:
@ijuma I had mistakenly thought it was taking a tuple there, and maybe that was just the way Scala converted a lambda to a BiConsumer, and the results had regressed back to close to what I was seeing with the iterator version.

When I ran it, it returned saw:
```
1140074.937 ± 75151.914    B/op
def foreachEntry(f: (K, V) => Unit): Unit = {
    pool.forEach((k, v) => f(k, v))
  }

1044120.722 ±  1676.137    B/op (included in an above comment)   
pool.forEach(new BiConsumer[K,V] {
      override def accept(t: K, u: V): Unit = f(t, u)
    })
```

Strangely I have just re-run the BiConsumer version and it returned `1188097.413 ±  1639.537    B/op`. I'm not sure why it's regressed from what I saw on a previous run. I am OK with using the version with a lambda if you are. I'm not sure I will have time to investigate it further, and it's still a good improvement compared to the version that used `foreach`.